source/_posts/2019-01-28-building-a-decentralized-reddit-with-embark-part-2.md

```diff
@@ -0,0 +1,194 @@
+
+Luckily wallet accounts can be easily accessed as they are emitted by Embark's `config()` function. All we have to do is attaching a resolution handler to `config()` and storing the emitted value:
+
+<pre class="javascript"><code>...
+let accounts = [];
+
+config({
+  contracts: {
+    DReddit: {}
+  }
+}, (err, _accounts) => {
+  accounts = _accounts;
+});</code></pre>
+
+Having that in place, our next test could look something like this:
+
+<pre class="javascript"><code>contract('DReddit', () => {
+  ...
+  it ('post should have correct data', async () => {
+    const post = await DReddit.methods.posts(postId).call();
+    assert.equal(web3.utils.toAscii(post.description), ipfsHash);
+    assert.equal(post.owner, accounts[0]);
+  });
+});</code></pre>
+
+## Testing `canVote()`
+
+Alright, next up let's quickly test if our `canVote()` method works the way as expected. As voting on posts that don't exist should never work, we will simply call `canVote()` on a post id that doesn't exist. This test is pretty straight forward:
+
+<pre class="javascript"><code>contract('DReddit', () => {
+  ...
+  it('should not be able to vote in an unexisting post', async () => {
+    const userCanVote = await DReddit.methods.canVote("123").call();
+    assert.equal(userCanVote, false);
+  });
+});</code></pre>
+
+We also want to make sure that `canVote()` resolves to `true` in case a user can indeed vote a certain post. We can again reuse the `postId` that we've stored earlier:
+
+<pre class="javascript"><code>it('should be able to vote in a post if account hasn\'t voted before', async () => {
+  const userCanVote = await DReddit.methods.canVote(postId).call();
+  assert.equal(userCanVote, true);
+});</code></pre>
+
+Wonderful, we have 5 passing tests now!
+
+## Testing `vote()`
+
+Of course we want to test whether one of our application's core features works as well. There's certainly different ways to verify whether `vote()` does what it's supposed to do, but for this tutorial we'll simply check whether the owner account of the vote emitted by the `NewVote` event is the same as the account that performed the vote. We can actually take some inspiration from our previous tests:
+
+<pre class="javascript"><code>contract('DReddit', () => {
+  ...
+  it("should be able to vote in a post", async () => {
+    const receipt = await DReddit.methods.vote(postId, 1).send();
+    const Vote = receipt.events.NewVote;
+    assert.equal(Vote.returnValues.owner, accounts[0]);
```

Comment:
Shall we explain that accounts[0] is the default account used.
Or make it explicit in the `send`?

Reply:
Yes I'll add a comment about that. Good point!